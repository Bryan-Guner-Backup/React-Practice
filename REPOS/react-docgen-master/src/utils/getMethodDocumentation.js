/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import { namedTypes as t } from 'ast-types';
import { getDocblock } from './docblock';
import getFlowType from './getFlowType';
import getTSType from './getTSType';
import getParameterName from './getParameterName';
import getPropertyName from './getPropertyName';
import getTypeAnnotation from './getTypeAnnotation';
import type { FlowTypeDescriptor, Importer } from '../types';
import resolveToValue from './resolveToValue';

type MethodParameter = {
  name: string,
  type?: ?FlowTypeDescriptor,
  optional?: boolean,
};

type MethodReturn = {
  type: ?FlowTypeDescriptor,
};

type MethodDocumentation = {
  name: string,
  docblock: ?string,
  modifiers: Array<string>,
  params: Array<MethodParameter>,
  returns: ?MethodReturn,
};

function getMethodFunctionExpression(methodPath, importer) {
  const exprPath = t.AssignmentExpression.check(methodPath.node)
    ? methodPath.get('right')
    : methodPath.get('value');
  return resolveToValue(exprPath, importer);
}

function getMethodParamsDoc(methodPath, importer) {
  const params = [];
  const functionExpression = getMethodFunctionExpression(methodPath, importer);

  // Extract param flow types.
  functionExpression.get('params').each(paramPath => {
    let type = null;
    const typePath = getTypeAnnotation(paramPath);
    if (typePath && t.Flow.check(typePath.node)) {
      type = getFlowType(typePath, null, importer);
      if (t.GenericTypeAnnotation.check(typePath.node)) {
        type.alias = typePath.node.id.name;
      }
    } else if (typePath) {
      type = getTSType(typePath, null, importer);
      if (t.TSTypeReference.check(typePath.node)) {
        type.alias = typePath.node.typeName.name;
      }
    }

    const param = {
      name: getParameterName(paramPath),
      optional: paramPath.node.optional,
      type,
    };

    params.push(param);
  });

  return params;
}

// Extract flow return type.
function getMethodReturnDoc(methodPath, importer) {
  const functionExpression = getMethodFunctionExpression(methodPath, importer);

  if (functionExpression.node.returnType) {
    const returnType = getTypeAnnotation(functionExpression.get('returnType'));
    if (returnType && t.Flow.check(returnType.node)) {
      return { type: getFlowType(returnType, null, importer) };
    } else if (returnType) {
      return { type: getTSType(returnType, null, importer) };
    }
  }

  return null;
}

function getMethodModifiers(methodPath) {
  if (t.AssignmentExpression.check(methodPath.node)) {
    return ['static'];
  }

  // Otherwise this is a method/property node

  const modifiers = [];

  if (methodPath.node.static) {
    modifiers.push('static');
  }

  if (methodPath.node.kind === 'get' || methodPath.node.kind === 'set') {
    modifiers.push(methodPath.node.kind);
  }

  const functionExpression = methodPath.get('value').node;
  if (functionExpression.generator) {
    modifiers.push('generator');
  }
  if (functionExpression.async) {
    modifiers.push('async');
  }

  return modifiers;
}

function getMethodName(methodPath, importer) {
  if (
    t.AssignmentExpression.check(methodPath.node) &&
    t.MemberExpression.check(methodPath.node.left)
  ) {
    const left = methodPath.node.left;
    const property = left.property;
    if (!left.computed) {
      return property.name;
    }
    if (t.Literal.check(property)) {
      return String(property.value);
    }
    return null;
  }
  return getPropertyName(methodPath, importer);
}

function getMethodAccessibility(methodPath) {
  if (t.AssignmentExpression.check(methodPath.node)) {
    return null;
  }

  // Otherwise this is a method/property node
  return methodPath.node.accessibility;
}

function getMethodDocblock(methodPath) {
  if (t.AssignmentExpression.check(methodPath.node)) {
    let path = methodPath;
    do {
      path = path.parent;
    } while (path && !t.ExpressionStatement.check(path.node));
    if (path) {
      return getDocblock(path);
    }
    return null;
  }

  // Otherwise this is a method/property node
  return getDocblock(methodPath);
}

// Gets the documentation object for a component method.
// Component methods may be represented as class/object method/property nodes
// or as assignment expresions of the form `Component.foo = function() {}`
export default function getMethodDocumentation(
  methodPath: NodePath,
  importer: Importer,
): ?MethodDocumentation {
  if (getMethodAccessibility(methodPath) === 'private') {
    return null;
  }

  const name = getMethodName(methodPath, importer);
  if (!name) return null;

  return {
    name,
    docblock: getMethodDocblock(methodPath),
    modifiers: getMethodModifiers(methodPath),
    params: getMethodParamsDoc(methodPath, importer),
    returns: getMethodReturnDoc(methodPath, importer),
  };
}
