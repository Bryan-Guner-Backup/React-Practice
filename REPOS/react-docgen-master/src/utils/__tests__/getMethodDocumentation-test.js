/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import {
  statement,
  noopImporter,
  makeMockImporter,
} from '../../../tests/utils';
import getMethodDocumentation from '../getMethodDocumentation';

describe('getMethodDocumentation', () => {
  const mockImporter = makeMockImporter({
    hello: statement(`
      export default () => {};
    `).get('declaration'),

    bar: statement(`
      export default (bar: number) => {};
    `).get('declaration'),

    baz: statement(`
      export default (): number => {};
    `).get('declaration'),
  });

  describe('name', () => {
    it('extracts the method name', () => {
      const def = statement(`
        class Foo {
          hello() {}
        }
      `);
      const method = def.get('body', 'body', 0);
      expect(getMethodDocumentation(method, noopImporter)).toEqual({
        name: 'hello',
        docblock: null,
        modifiers: [],
        returns: null,
        params: [],
      });
    });

    it('handles function assignment', () => {
      const def = statement(`
        class Foo {
          hello = () => {}
        }
      `);
      const method = def.get('body', 'body', 0);
      expect(getMethodDocumentation(method, noopImporter)).toEqual({
        name: 'hello',
        docblock: null,
        modifiers: [],
        returns: null,
        params: [],
      });
    });

    it('handles computed method name', () => {
      const def = statement(`
        class Foo {
          [foo]() {}
        }
      `);
      const method = def.get('body', 'body', 0);
      expect(getMethodDocumentation(method, noopImporter)).toMatchSnapshot();
    });

    it('ignores complex computed method name', () => {
      const def = statement(`
        class Foo {
          [() => {}]() {}
        }
      `);
      const method = def.get('body', 'body', 0);
      expect(getMethodDocumentation(method, noopImporter)).toMatchSnapshot();
    });

    it('resolves assignment of imported function', () => {
      const def = statement(`
        class Foo {
          hello = hello;
        }
        import hello from 'hello';
      `);
      const method = def.get('body', 'body', 0);
      expect(getMethodDocumentation(method, mockImporter)).toEqual({
        name: 'hello',
        docblock: null,
        modifiers: [],
        returns: null,
        params: [],
      });
    });
  });

  describe('docblock', () => {
    it('extracts the method docblock', () => {
      const def = statement(`
        class Foo {
          /**
           * Don't use this!
           */
          foo() {}
        }
      `);
      const method = def.get('body', 'body', 0);
      expect(getMethodDocumentation(method, noopImporter)).toEqual({
        name: 'foo',
        docblock: "Don't use this!",
        modifiers: [],
        returns: null,
        params: [],
      });
    });

    it('extracts docblock on function assignment', () => {
      const def = statement(`
        class Foo {
          /**
           * Don't use this!
           */
          foo = () => {}
        }
      `);
      const method = def.get('body', 'body', 0);
      expect(getMethodDocumentation(method, noopImporter)).toEqual({
        name: 'foo',
        docblock: "Don't use this!",
        modifiers: [],
        returns: null,
        params: [],
      });
    });
  });

  describe('parameters', () => {
    function methodParametersDoc(params) {
      return {
        name: 'foo',
        docblock: null,
        modifiers: [],
        returns: null,
        params,
      };
    }

    it('extracts flow type info', () => {
      const def = statement(`
        class Foo {
          foo(bar: number) {}
        }
      `);
      const method = def.get('body', 'body', 0);
      expect(getMethodDocumentation(method, noopImporter)).toEqual(
        methodParametersDoc([
          {
            name: 'bar',
            type: { name: 'number' },
          },
        ]),
      );
    });

    it('extracts flow type info on function assignment', () => {
      const def = statement(`
        class Foo {
          foo = (bar: number) => {}
        }
      `);
      const method = def.get('body', 'body', 0);
      expect(getMethodDocumentation(method, noopImporter)).toEqual(
        methodParametersDoc([
          {
            name: 'bar',
            type: { name: 'number' },
          },
        ]),
      );
    });

    it('resolves flow type info on imported functions', () => {
      const def = statement(`
        class Foo {
          foo = bar
        }
        import bar from 'bar';
      `);
      const method = def.get('body', 'body', 0);
      expect(getMethodDocumentation(method, mockImporter)).toEqual(
        methodParametersDoc([
          {
            name: 'bar',
            type: { name: 'number' },
          },
        ]),
      );
    });

    describe('modifiers', () => {
      function methodModifiersDoc(modifiers) {
        return {
          name: 'foo',
          docblock: null,
          modifiers,
          returns: null,
          params: [],
        };
      }

      it('detects no modifiers', () => {
        const def = statement(`
          class Foo {
            foo() {}
          }
        `);
        const method = def.get('body', 'body', 0);
        expect(getMethodDocumentation(method, noopImporter)).toEqual(
          methodModifiersDoc([]),
        );
      });

      it('detects static functions', () => {
        const def = statement(`
          class Foo {
            static foo() {}
          }
        `);
        const method = def.get('body', 'body', 0);
        expect(getMethodDocumentation(method, noopImporter)).toEqual(
          methodModifiersDoc(['static']),
        );
      });

      it('detects generators', () => {
        const def = statement(`
          class Foo {
            *foo () {}
          }
        `);
        const method = def.get('body', 'body', 0);
        expect(getMethodDocumentation(method, noopImporter)).toEqual(
          methodModifiersDoc(['generator']),
        );
      });

      it('detects async functions', () => {
        const def = statement(`
          class Foo {
            async foo () {}
          }
        `);
        const method = def.get('body', 'body', 0);
        expect(getMethodDocumentation(method, noopImporter)).toEqual(
          methodModifiersDoc(['async']),
        );
      });

      it('detects static async functions', () => {
        const def = statement(`
          class Foo {
            static async foo () {}
          }
        `);
        const method = def.get('body', 'body', 0);
        expect(getMethodDocumentation(method, noopImporter)).toEqual(
          methodModifiersDoc(['static', 'async']),
        );
      });
    });

    describe('returns', () => {
      function methodReturnDoc(returnValue) {
        return {
          name: 'foo',
          docblock: null,
          modifiers: [],
          returns: returnValue,
          params: [],
        };
      }

      it('returns null if return is not documented', () => {
        const def = statement(`
          class Foo {
            foo () {}
          }
        `);
        const method = def.get('body', 'body', 0);
        expect(getMethodDocumentation(method, noopImporter)).toEqual(
          methodReturnDoc(null),
        );
      });

      it('extracts flow types', () => {
        const def = statement(`
          class Foo {
            foo (): number {}
          }
        `);
        const method = def.get('body', 'body', 0);
        expect(getMethodDocumentation(method, noopImporter)).toEqual(
          methodReturnDoc({
            type: { name: 'number' },
          }),
        );
      });

      it('extracts flow types on function assignment', () => {
        const def = statement(`
          class Foo {
            foo = (): number => {}
          }
        `);
        const method = def.get('body', 'body', 0);
        expect(getMethodDocumentation(method, noopImporter)).toEqual(
          methodReturnDoc({
            type: { name: 'number' },
          }),
        );
      });

      it('resolves flow types on imported functions', () => {
        const def = statement(`
          class Foo {
            foo = baz
          }
          import baz from 'baz';
        `);
        const method = def.get('body', 'body', 0);
        expect(getMethodDocumentation(method, mockImporter)).toEqual(
          methodReturnDoc({
            type: { name: 'number' },
          }),
        );
      });
    });

    describe('private', () => {
      it('ignores private typescript methods', () => {
        const def = statement(
          `
          class Foo {
            private foo() {}
          }
        `,
          { parserOptions: { plugins: ['typescript'] } },
        );
        const method = def.get('body', 'body', 0);
        expect(getMethodDocumentation(method, noopImporter)).toMatchSnapshot();
      });

      it.skip('ignores private methods', () => {
        const def = statement(
          `
          class Foo {
            #foo() {}
          }
        `,
          { parserOptions: { plugins: ['classPrivateMethods'] } },
        );
        const method = def.get('body', 'body', 0);
        expect(getMethodDocumentation(method, noopImporter)).toMatchSnapshot();
      });
    });
  });
});
