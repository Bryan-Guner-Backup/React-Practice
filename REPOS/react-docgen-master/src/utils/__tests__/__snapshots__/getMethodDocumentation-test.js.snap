// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMethodDocumentation name handles computed method name 1`] = `
Object {
  "docblock": null,
  "modifiers": Array [],
  "name": "@computed#foo",
  "params": Array [],
  "returns": null,
}
`;

exports[`getMethodDocumentation name ignores complex computed method name 1`] = `null`;

exports[`getMethodDocumentation parameters private ignores private typescript methods 1`] = `null`;
